python/semantic_kernel/connectors/ai/open_ai/services/realtime/open_ai_realtime_base.py

```diff
@@ -0,0 +1,201 @@
+                self._update_function_choice_settings_callback(),
+                kernel=self.kernel,  # type: ignore
+            )
+            await self.send(SendEvents.SESSION_UPDATE, settings=self._current_settings)
+        if chat_history and len(chat_history) > 0:
+            for msg in chat_history.messages:
+                await self.send(SendEvents.CONVERSATION_ITEM_CREATE, item=msg)
+        if create_response:
+            await self.send(SendEvents.RESPONSE_CREATE)
+
+    @override
+    def get_prompt_execution_settings_class(self) -> type["PromptExecutionSettings"]:
+        from semantic_kernel.connectors.ai.open_ai.prompt_execution_settings.open_ai_realtime_execution_settings import (  # noqa
+            OpenAIRealtimeExecutionSettings,
+        )
+
+        return OpenAIRealtimeExecutionSettings
+
+    @override
+    def _update_function_choice_settings_callback(
+        self,
+    ) -> Callable[[FunctionCallChoiceConfiguration, "PromptExecutionSettings", FunctionChoiceType], None]:
+        return update_settings_from_function_call_configuration
+
+    async def _handle_function_call_arguments_done(
+        self,
+        event: ResponseFunctionCallArgumentsDoneEvent,
+    ) -> None:
+        """Handle response function call done."""
+        if not self.kernel or (
+            self._current_settings
+            and self._current_settings.function_choice_behavior
+            and not self._current_settings.function_choice_behavior.auto_invoke_kernel_functions
+        ):
+            return
+        plugin_name, function_name = self._call_id_to_function_map.pop(event.call_id, "-").split("-", 1)
+        if not plugin_name or not function_name:
+            logger.error("Function call needs to have a plugin name and function name")
+            return
+        item = FunctionCallContent(
+            id=event.item_id,
+            plugin_name=plugin_name,
+            function_name=function_name,
+            arguments=event.arguments,
+            index=event.output_index,
+            metadata={"call_id": event.call_id},
+        )
+        chat_history = ChatHistory()
+        await self.kernel.invoke_function_call(item, chat_history)
+        created_output = chat_history.messages[-1]
+        # This returns the output to the service
+        await self.send(SendEvents.CONVERSATION_ITEM_CREATE, item=created_output)
+        # The model doesn't start responding to the tool call automatically, so triggering it here.
+        await self.send(SendEvents.RESPONSE_CREATE)
+        # This allows a user to have a full conversation in his code
+        await self.receive_buffer.put((ListenEvents.RESPONSE_FUNCTION_CALL_ARGUMENTS_DONE, created_output))
+
+    @override
+    async def start_listening(
+        self, settings: PromptExecutionSettings | None = None, chat_history: ChatHistory | None = None, **kwargs: Any
+    ) -> None:
+        pass
+
+    @override
+    async def start_sending(self, **kwargs: Any) -> None:
```

Comment:
In the `start_sending` child classes: is there anything we'd need to add to better handle shutdowns? or when there is no more data to handle?

Reply:
both sending and listening are loops that execute until they are stopped and both just react to things coming in, we could look at a bit nicer propagation of a session close, but for sending that would just mean the queue is empty, but it can be empty during the session, while the service keeps sending until we call close, so we don't then want to keep going either (I think)